### ISSUE TITLE
Tuple type serialisation

### ISSUE THREAD

**Issue by mattcorner** [remix-run/remix#6566]
### What version of Remix are you using?

1.16.1

### Are all your remix dependencies & dev-dependencies using the same version?

- [X] Yes

### Steps to Reproduce

We often have typings that narrow down, so a top level type might define a property as `string[]` but a refined version of that type might narrow it down to `["Some value"]`. This works fine, until Remix infers and serialises the type across the network boundary. 

I've taken the Serialization code from the repository, and generated a Typescript playground showing the issue. You'll see that the tuple gets interpretted as an array and has all of its keys incorrectly defined.

[Typescript playground](https://www.typescriptlang.org/play?#code/PTAEDMHsCcGMFNQBUCeAHeBnW0CWaAXUAyUTXAWzQBtdwViALRA9eAKFY1AAVp4CrADxIAfKAC8oAN6gA2gGlQuAHagA1vBSRwyALoAuZIr2gAvqABkMswG52nNqABSmSCr6VcBXADdEEuygoAA+ZAR4KgDmQaGgKgCuFABG8NCxYcmQkNTwAIYqGaAAyhGqMcFhAHJJqemVoABC2bkFRYnU1PZciFXuru6eFN5+AaAJKgAm8OCq8JNxAGITsD7ucZgoKTn27MAAVEH7oAAGBSgnyphMiO7UDD1MeUSwBaC5RNoJoPAAjgnPRAnAAMlwA7t5GKcAIwnI4leCIRiCNCYAwgTAEPKwdSQfzQcDUSBggB0sEgFGAeWAABYAJx0gBMAA4AMwANmBwGh9OBwLp0KOwEc3AAkpgAIIqFAicRSYE-AAeBHgU2u0KsyFAAH5iNAEogjOA8tRMPBdiBQGh+IJcGkALS4KIqGAcR7FNK4E24ABe8Flkl4NuEsXFUplYiVKrVeoNOtA51ABliSCjqsm1wGHjwwx8-njqeTwVT8GV6eufRUWaGI3zupU8HxSZTaZjshIzmKAHkqgAKACURlU4DSoAAquZ4xOi1rS9GM-JTLq5KYZyWyzG5BN1C6wSoADSgEnH7e7lQrpcIvDev1IBI0f2RtethcAJXykzuKAl0GgeRlw6jmO4i6r2E5zuWoCVtWOa1og9YJJ0SZXl6tB+kIwH9iuzbFi+1yQMkABW8CrPGHrXmh8BdkRJEEBhUwzHMkxIJAXaELg7gmrK4gzg2+LsKIFr7McnY9mQno3ogcgEPeuSYHovbIgQqLosAYLqSSPTYHghDUAUUQkjAUTAJ+sCYMAjAFJMWSQOowCMsABHEasmAksiFDUAAxDJD72lp-agPswruhJlF3g+Ih4fIp7EgeR4nioO6xRecoyLEcjqMoaiaNouhIIYxjqKYEExtBbjZl4ebwfEiHUMh5Gob6j4ZXognsHYDgHCJ3ZVOJFFNaATm0dcvYukQqiwNQCSTOUCa-v+mABVZoCTXkmBmtcQUiogDWSdRzl0eu874TRqypdI6WZaoGhaDoWprYVxUbguZWDLBVXxnxo5GEVRi7ZRIgtW1HV7GA1oCD4DpOi6-DbShkkACIzGk-CTJFJVviRMBo5ikRRIeMV7qI52XVlN25fdmCxMWLVRXwFK4GaQiEyoIEaDhwSPVFr0VbmoyfY233UxosSGMLSC0xj1z08MTOAdA46iMLuoy4z-r-U1GHE1FX0K7qABEYIEOA+vIarTMa+hwFK5z9WhZrEtFeIOuC3r7N-fb6GO619ggwc8KLDACZqENpFIIedDKEQlnXIm1qQBg0CQ9cBCWUQeT8Am4wqBxagQqnpwTNMswNpMJz7vCFB5JoTCQGaWUkIN7GcXV8eJ8nWWYh+JJHPCACiip5FQuRGLIeRGLjs1hEXjGlxY9r2uIY-ahPZTRGYQpw2ODEl-MLFsWsKhcUdkGhwQJPBJa9PtwwjPxJA6fZ7noD51CJwz7vZeHpo8BoAmmCOipsEDKZMcp3VTA9D+TEorewFk2H6BVva+01BdS+YBr5pFYKAAA+ozbBWcJjP1fqcYuTETgkgROnfCzcj51V7L4L0pxtQnCWlMUA-AKB4iBFA0ulxwDQApDcJ+7hSbXTAXlf+2dSGlxgbTXUmUjC6z0CvUAA9JrTX9OLFqFdbY8PmLEYGuxHg-mSN4P80AUDLBUKsZ+UhezHhJBnKIaJg4oBXAFCQ4gWZGKcAAWRQOFXIVQM4CLBPMQMk9ojYWsHIfWxQKSIF8CaA0+s9DdD8QE2S8AeB6WulIWJ8SKCJOSfAVJDgQr9T9JMfxgT4DBN-MScJUhLb+hqVk+poT5hKwqY1KpbSHw5LyHk+GAN+m5EGaoJW7ByQqExKACgmSHwdMaZMD2lT5hjLqSElZgYCkJNAEkqapS9DTPcHMhZtSJkqDWb0jZizxm5LUPkuJ+zDkpL0EAA)

### Expected Behavior

The serialised type to remain `string[] & ["Some value"]`

### Actual Behavior

The seralised type becomes 

```
{
    [x: number]: "Some value";
    length: "Some value";
    toString: null;
    toLocaleString: null;
    pop: null;
    push: null;
    concat: null;
    join: null;
    reverse: null;
    shift: null;
    slice: null;
    sort: null;
    splice: null;
    ... 20 more ...;
    0: "Some value";
}
```

**Comment by pcattori** [comment]
@mattcorner could you help me understand why `string[] & ["Some value"]` is a useful type for you? Why not narrow down to `["Some value"]`?

**Comment by MichaelDeBoey** [comment]
Fixed by #6616
